cpp/dolfinx/la/MatrixCSR.h

```diff
@@ -0,0 +1,491 @@
+      int row_id = local_size[0] + i;
+      for (int j = _row_ptr[row_id]; j < _row_ptr[row_id + 1]; ++j)
+      {
+        // Get index position in send buffer
+        const std::int32_t idx_pos = insert_pos[neighbor_rank];
+
+        // Pack send data (row, col) as global indices
+        ghost_index_data[idx_pos] = ghosts0[i];
+        if (const std::int32_t col_local = _cols[j]; col_local < local_size[1])
+          ghost_index_data[idx_pos + 1] = col_local + local_range[1][0];
+        else
+          ghost_index_data[idx_pos + 1] = ghosts1[col_local - local_size[1]];
+
+        insert_pos[neighbor_rank] += 2;
+      }
+    }
+
+    // Create and communicate adjacency list to neighborhood
+    const graph::AdjacencyList<std::int64_t> ghost_index_data_out(
+        std::move(ghost_index_data), std::move(index_send_disp));
+    const graph::AdjacencyList<std::int64_t> ghost_index_data_in
+        = dolfinx::MPI::neighbor_all_to_all(_comm.comm(), ghost_index_data_out);
+
+    // Store received offsets for future use, when transferring data values.
+    _val_recv_disp.resize(ghost_index_data_in.offsets().size());
+    std::transform(ghost_index_data_in.offsets().begin(),
+                   ghost_index_data_in.offsets().end(), _val_recv_disp.begin(),
+                   [](int d) { return d / 2; });
+
+    // Global to local map for ghost columns
+    std::map<std::int64_t, std::int32_t> global_to_local;
+    std::int32_t local_i = local_size[1];
+    for (std::int64_t global_i : ghosts1)
+      global_to_local.insert({global_i, local_i++});
+
+    // Compute location in which data for each index should be stored
+    // when received
+    const std::vector<std::int64_t>& ghost_index_array
+        = ghost_index_data_in.array();
+    for (std::size_t i = 0; i < ghost_index_array.size(); i += 2)
+    {
+      // Row must be on this process
+      const std::int32_t local_row = ghost_index_array[i] - local_range[0][0];
+      assert(local_row >= 0 and local_row < local_size[0]);
+
+      // Column may be owned or unowned
+      std::int32_t local_col = ghost_index_array[i + 1] - local_range[1][0];
+      if (local_col < 0 or local_col >= local_size[1])
+      {
+        const auto it = global_to_local.find(ghost_index_array[i + 1]);
+        assert(it != global_to_local.end());
+        local_col = it->second;
+      }
+      auto cit0 = std::next(_cols.begin(), _row_ptr[local_row]);
+      auto cit1 = std::next(_cols.begin(), _row_ptr[local_row + 1]);
+
+      // Find position of column index and insert data
+      auto cit = std::lower_bound(cit0, cit1, local_col);
+      assert(cit != cit1);
+      assert(*cit == local_col);
+      std::size_t d = std::distance(_cols.begin(), cit);
+      _unpack_pos.push_back(d);
+    }
+  }
+
+  /// Move constructor
+  /// @todo Check handling of MPI_Request
+  MatrixCSR(MatrixCSR&& A) = default;
+
+  /// Set all non-zero local entries to a value
```

Comment:
Should it also set ghost rows?

Reply:
Yes, I agree. Does it make sense to 'set to a value' or is it always just going to be zero?

Reply:
I think it will almost always be zero, but it doesn't introduce any complexity allowing to be some other value. I can image maybe creating an identity matrix for testing.

Reply:
OK, but you need to be careful with that, if multiple entries exist on different processes for the same global location.